woocommerce/woocommerce-ios
# Update analytics to show the latest time range based on site time zone instead of device time zone

Part of #10842


## Why


As clarified in https://github.com/woocommerce/woocommerce-android/issues/9887#issuecomment-1855990495, we'd like to use the store time zone to calculate the latest time range like today/this week/this month/this year for scenarios when the store time zone and device time zone are on different dates. I also opened a discussion in pe5pgL-4bG-p2 about this, but wanted to open the PR in any case because the changes enable the app to switch the latest time range time zone base with just 2 lines of code.

Potential future task: update the order dates to also be in the store time zone, based on the decision from the discussion

## How

The main changes were to add a time zone parameter to 2 of the 4 analytics API requests, `OrderStatsRemoteV4 .loadOrderStats` and `ProductsReportsRemote.loadTopProductsReport`. The other 2 API requests on the WPCOM site visit stats and summary already use the store time zone for the time range `Date` -> `String` conversion. Each API request includes at least one date string without a time zone like `2023-12-16T23:59:59`, and the backend loads the stats using the date string(s) in the store time zone. The addition of the time zone parameter requires DI'ing the time zone to 3 Yosemite actions `StatsActionV4.retrieveStats`, `StatsActionV4.retrieveCustomStats`, and `StatsActionV4.retrieveTopEarnerStats`. The store just passes the new parameter to the corresponding networking call.

Then, the change to sync stats with the latest time range based on the store time zone instead of the device time zone was in `StoreStatsAndTopPerformersViewController`.

From this change, there were 2 issues where the UI wasn't showing the data on the day in the store time zone:

- The top performers data weren't showing the data on the day in the store time zone when the store time zone and device time zone were on different days
  - The reason was because the top performers view controllers were initialized with the device time zone, then the time zone wasn't updated when the dashboard view controller syncs data with the store time zone later on. The time zone could also change when the user pulls to refresh, if the site time zone was updated earlier
  - The fix was in https://github.com/woocommerce/woocommerce-ios/pull/11479/commits/4b535063d322740db8e41913dd9e9a00210e63b9 to update the results controller when the time zone changes. Originally I also tried refactoring this to use Combine so that it's easier to read, but the results controller's `onDidChangeContent` and `onDidResetContent` callbacks made it hard to change the data flow
- The analytics hub title wasn't showing the date in the store zone
  - It was because `AnalyticsHubTimeRangeSelection` was using the device time zone as the default time zone
  - The fix was in https://github.com/woocommerce/woocommerce-ios/pull/11479/commits/4eff5734c2c2cdcfd464587b025140d167a47a67

While debugging the top performers code, I also noticed `TopPerformersSectionHeaderView` wasn't used anymore and deleted this file. This view was replaced by `TopPerformersView` shared in both the dashboard and analytics hub.

## Testing instructions


Prerequisite: at the time of testing, the device time zone should be on a different day from the store time zone

- Go to the orders tab, create an order with at least one product and mark it as paid like by tapping `Collect Payment` > `Cash` > `Mark as Paid`
- Go to the my store tab, tap on the `Today` tab then pull to refresh (if the new order isn't included in the analytics, try pulling to refresh in about half a minute) --> "today" should be the day in the store time zone and not the device time zone. the new order & products should be reflected in the stats in the today tab

## Screenshots


\ | before | after
-- | -- | --
today stats | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/c18027f7-1a82-4fa8-be3f-4f0e2194121d" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/0b0122f4-f08f-459b-9a12-85c1ecf5528f" width="300" />
analytics hub | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/23008616-2a06-4a17-8f84-1be3e9502fe7" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/ae4afa13-63a2-4d5f-8370-1b6a2d1be5ab" width="300" />

---
